modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceNodeConverter.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.talend.avro;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Field;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.IndexedRecord;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.talend.daikon.avro.NameUtil;
+
+/**
+ * @author Zoltán Takács
+ */
+@SuppressWarnings("rawtypes")
+public class ResourceNodeConverter
+	extends BaseConverter<JsonNode, IndexedRecord> {
+
+	public ResourceNodeConverter(Schema schema) {
+		super(JsonNode.class, schema);
+
+		initConverters(schema);
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public IndexedRecord convertToAvro(JsonNode resource) {
+		IndexedRecord record = new GenericData.Record(getSchema());
+
+		Iterator<Map.Entry<String, JsonNode>> jsonFields = resource.fields();
+
+		// Already used names for the fields
+
+		Set<String> normalizedJsonFiledNames = new HashSet<>();
+		int i = 0;
+		int pos = -1;
+
+		while (jsonFields.hasNext()) {
+			Map.Entry<String, JsonNode> field = jsonFields.next();
+
+			String fieldName = NameUtil.correct(
+				field.getKey(), i, normalizedJsonFiledNames);
+
+			normalizedJsonFiledNames.add(fieldName);
+
+			i++;
+
+			for (Field schemaField : schemaFields) {
+				if (fieldName.equals(schemaField.name())) {
+					pos = schemaField.pos();
+
+					break;
+				}
+			}
+
+			if (pos >= 0) {
+				JsonNode resourceJsonNode = field.getValue();
+
+				Object value = avroConverters[pos].convertToAvro(
+					resourceJsonNode.asText());
+
+				record.put(pos, value);
+				pos = -1;
+			}
+			else {
+				if ("_type".equals(fieldName) || "_context".equals(fieldName)) {
```

Comment:
All these nested condition blocks look a little weird, wouldn't it better to just have a single `else if` with:
```java
else if (_log.isDebugEnabled() && !fieldName.equals("_type") && !fieldName.equals("_context")) {
```